### PR TITLE
Rosetta Scaffolding

### PR DESCRIPTION
**Pull request by fosskers**
This PR lays the groundwork for our support of the Rosetta API.

### TODO

- [x] Basic API layout
- [x] `SomeServer` plumbing
- [x] Associate endpoints with real types
- [x] Update nix and cabal settings to correct `rosetta` pin

### CODE REVIEW

File: src/Chainweb/Chainweb.hs [hunk 1]
```diff
@@ -374,6 +375,7 @@ instance ToJSON ChainwebConfiguration where
         , "reorgLimit" .= _configReorgLimit o
         , "validateHashesOnReplay" .= _configValidateHashesOnReplay o
         , "allowReadsInLocal" .= _configAllowReadsInLocal o
+        -- , "rosetta" .= _configRosetta o
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs:378]: Disabled for now.

File: src/Chainweb/Chainweb.hs [hunk 2]
```diff
@@ -442,7 +448,6 @@ data Chainweb logger cas = Chainweb
     , _chainwebCutResources :: !(CutResources logger cas)
     , _chainwebMiner :: !(Maybe (MinerResources logger cas))
     , _chainwebCoordinator :: !(Maybe (MiningCoordination logger cas))
-    , _chainwebHeaderStream :: !HeaderStream
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs:445]: This is a redundant field - the setting itself is already contained in `ChainwebConfiguration`.

File: src/Chainweb/Rosetta/RestAPI.hs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    :<|> "rosetta" :> "network" :> "status"
+        :> ReqBody '[JSON] NetworkReq
+        :> Post '[JSON] NetworkStatusResp
+
+type RosettaApi (v :: ChainwebVersionT) = 'ChainwebEndpoint v :> Reassoc RosettaApi_
+
+data RosettaFailure
+    = RosettaChainUnspecified
+    | RosettaInvalidChain Text
+    | RosettaMempoolBadTx
```
Comment (fosskers) [inline src/Chainweb/Rosetta/RestAPI.hs:83]: This ADT will be extended as new errors arise.

File: src/Chainweb/Rosetta/RestAPI.hs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+rosettaError :: RosettaFailure -> RosettaError
+rosettaError RosettaChainUnspecified = RosettaError 0 "No SubNetwork (chain) specified" False
+rosettaError (RosettaInvalidChain cid) = RosettaError 1 ("Invalid chain value: " <> cid) False
```
Comment (fosskers) [inline src/Chainweb/Rosetta/RestAPI.hs:87]: It's unclear whether we're allowed to have "generalized" error messages like this,  or whether all messages must be static.

File: src/Chainweb/BlockHeaderDB/RestAPI.hs [hunk 5]
```diff
@@ -357,8 +356,6 @@ someBlockHeaderDbApis v = mconcat . fmap (someBlockHeaderDbApi v)
 -- -------------------------------------------------------------------------- --
 -- BlockHeader Event Stream
 
-newtype HeaderStream = HeaderStream Bool
```
Comment (LindaOrtega) [inline src/Chainweb/BlockHeaderDB/RestAPI.hs:360]: Why was this deleted?

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import Chainweb.Rosetta.RestAPI
+import Chainweb.Version
+
+---
+
+rosettaServer
+    :: forall a (v :: ChainwebVersionT)
+    . [(ChainId, MempoolBackend a)]
+    -> Server (RosettaApi v)
+rosettaServer ms = (const $ error "not yet implemented")
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:48]: So if we need more backend components we add it to the list here?

Comment (LindaOrtega) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:48]: So this handlers are in order based on https://github.com/kadena-io/chainweb-node/pull/1021/files#diff-b76452fda2871d503bcad2b60f35b6f5R41?
So this initial `(const $ error "not yet implemented")` corresponds to the account balance endpoint?

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+-- Construction Handlers
+
+constructionMetadataH :: ConstructionMetadataReq -> Handler ConstructionMetadataResp
+constructionMetadataH _ = error "not yet implemented"
+
+--------------------------------------------------------------------------------
+-- Mempool Handlers
+
+mempoolH :: MempoolReq -> [(ChainId, MempoolBackend a)] -> Handler MempoolResp
+mempoolH (MempoolReq (NetworkId _ _ msni)) ms = case msni of
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:86]:
```suggestion
mempoolH (MempoolReq (NetworkId _ _ subNet)) ms = case subNet of
```

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    Nothing -> throwRosetta RosettaChainUnspecified
+    Just (SubNetworkId n _) ->
+        case readMaybe @ChainId (T.unpack n) >>= flip lookup ms of
+            Nothing -> throwRosetta $ RosettaInvalidChain n
+            Just _ -> do
+                error "not yet implemented"  -- TODO!
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:92]:
```suggestion
    Just (SubNetworkId cid _) ->
        case readMaybe @ChainId (T.unpack cid) >>= flip lookup ms of
            Nothing -> throwRosetta $ RosettaInvalidChain cid
            Just _ -> do
                error "not yet implemented"  -- TODO!
```

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+mempoolTransactionH
+    :: MempoolTransactionReq
+    -> [(ChainId, MempoolBackend a)]
+    -> Handler MempoolTransactionResp
+mempoolTransactionH mtr ms = runExceptT work >>= either throwRosetta pure
+  where
+    MempoolTransactionReq (NetworkId _ _ msni) (TransactionId ti) = mtr
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:100]: `msni` -- same naming suggestion as in previous comment

File: src/Chainweb/BlockHeaderDB/RestAPI.hs (hunk 5 above)
Comment (fosskers) [reply src/Chainweb/BlockHeaderDB/RestAPI.hs:360]: I mentioned it in one of my own comments.

File: src/Chainweb/Rosetta/RestAPI/Server.hs (hunk 6 above)
Comment (fosskers) [reply src/Chainweb/Rosetta/RestAPI/Server.hs:48]: Yes, alongside the `ms`.

Comment (fosskers) [reply src/Chainweb/Rosetta/RestAPI/Server.hs:48]: Exactly.

File: src/Chainweb/Rosetta/RestAPI/Server.hs (hunk 8 above)
Comment (fosskers) [reply src/Chainweb/Rosetta/RestAPI/Server.hs:92]: This handler is a stub anyway - we can update any aspect of it later in follow-up PRs.

File: src/Chainweb/Rosetta/RestAPI/Server.hs (hunk 9 above)
Comment (fosskers) [reply src/Chainweb/Rosetta/RestAPI/Server.hs:100]: "Maybe Sub Net Identifier" was the motivation. It's only used in one place, so the name isn't as important.
